lib/model.js

```diff
@@ -458,149 +452,158 @@ function addAuthToUrl({ isPasswordProtected }) {
-            if (this.readPreferenceTags) {
-              req.query.readPreferenceTags = Object.keys(
-                this.readPreferenceTags
-              )
-                .map(tag => `${tag}:${this.readPreferenceTags[tag]}`)
-                .join(',');
-            }
-          } else if (this[item] !== '') {
-            req.query[item] = this[item];
-          }
+  Object.keys(CONNECTION_STRING_OPTIONS).forEach(item => {
+    if (typeof model[item] !== 'undefined' && !req.query[item]) {
+      if (item === 'compression') {
+        if (model.compression && model.compression.compressors) {
+          req.query.compressors = model.compression.compressors.join(',');
         }
-      });
 
-      if (this.ssl) {
-        req.query.ssl = this.ssl;
-      } else if (
-        includes(['UNVALIDATED', 'SYSTEMCA', 'SERVER', 'ALL'], this.sslMethod)
-      ) {
-        req.query.ssl = 'true';
-      } else if (this.sslMethod === 'IFAVAILABLE') {
-        req.query.ssl = 'prefer';
-      } else if (this.sslMethod === 'NONE') {
-        req.query.ssl = 'false';
+        if (model.compression && model.compression.zlibCompressionLevel) {
+          req.query.zlibCompressionLevel = model.compression.zlibCompressionLevel;
+        }
+      } else if (item === 'authMechanismProperties') {
+        if (model.authMechanismProperties) {
+          req.query.authMechanismProperties = Object.keys(
+            model.authMechanismProperties
+          )
+            .map(tag => `${tag}:${model.authMechanismProperties[tag]}`)
+            .join(',');
+        }
+      } else if (item === 'readPreferenceTags') {
+        if (model.readPreferenceTags) {
+          req.query.readPreferenceTags = Object.keys(
+            model.readPreferenceTags
+          )
+            .map(tag => `${tag}:${model.readPreferenceTags[tag]}`)
+            .join(',');
+        }
+      } else if (model[item] !== '') {
+        req.query[item] = model[item];
       }
-
-      return toURL(req);
     }
-  },
+  });
+
+  if (model.ssl) {
+    req.query.ssl = model.ssl;
+  } else if (
+    includes(['UNVALIDATED', 'SYSTEMCA', 'SERVER', 'ALL'], model.sslMethod)
+  ) {
+    req.query.ssl = 'true';
+  } else if (model.sslMethod === 'IFAVAILABLE') {
+    req.query.ssl = 'prefer';
+  } else if (model.sslMethod === 'NONE') {
+    req.query.ssl = 'false';
+  }
+
+  return req;
+};
+
+assign(derived, {
   safeUrl: {
     cache: false,
     fn() {
-      return addAuthToUrl.call(this, { isPasswordProtected: true });
+      return addAuthToUrl.call(this, { url: toURL(prepareRequest(this)), isPasswordProtected: true });
     }
   },
   driverUrl: {
     cache: false,
     fn() {
-      return addAuthToUrl.call(this, { isPasswordProtected: false });
+      return addAuthToUrl.call(this, { url: toURL(prepareRequest(this)), isPasswordProtected: false });
+    }
+  },
+  driverUrlWithSsh: {
+    cache: false,
+    fn() {
+      const req = clone(prepareRequest(this));
```

Comment:
Before I think we were using `cloneDeep` in order to make sure we are not setting any nested properties - is this something we don't need to use anymore? Setting something on `req.query` would mutate the original model here.

Reply:
We used `cloneDeep` because we cloned the entire connection model that had `driverOptions` property first and then nested ssh properties. Currently, we do not set nested properties here, only 2 ssh fields, therefore we can use just clone, but later if we want to update the logic and add deeper nested properties we can replace it with `cloneDeep`.